File: crates/1bit-power/src/metrics.rs
// One-line JSON metrics sampler for `halo-power log`.
//
// Reads `/sys/class/hwmon` entries (temp_input, power1_average, freq) and
// emits a single serde_json line. Deliberately tolerant: anything missing
// just becomes `null`. A systemd timer fires this every 30 s and pipes to
// journald where Loki/`halo logs` can scoop it up.

use anyhow::Result;
use serde::Serialize;
use std::fs;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Serialize)]
pub struct Sample {
    pub ts_unix: u64,
    pub host: String,
    pub tctl_c:      Option<f32>,
    pub edge_c:      Option<f32>,
    pub pkg_power_w: Option<f32>,
    pub ec_temp_c:   Option<i32>,
    pub ec_power_mode: Option<String>,
    pub ec_fan1_rpm: Option<u32>,
    pub ec_fan2_rpm: Option<u32>,
    pub ec_fan3_rpm: Option<u32>,
}

pub fn sample() -> Result<String> {
    let ec = crate::ec::EcBackend::new();
    let (temp, mode, rpm1, rpm2, rpm3) = if ec.available() {
        (
            ec.temp_c().ok(),
            ec.power_mode().ok(),
            ec.fan(1).ok().map(|f| f.rpm),
            ec.fan(2).ok().map(|f| f.rpm),
            ec.fan(3).ok().map(|f| f.rpm),
        )
    } else {
        (None, None, None, None, None)
    };
    let s = Sample {
        ts_unix: SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_secs()).unwrap_or(0),
        host: hostname(),
        tctl_c:      read_hwmon("k10temp", "temp1_input").map(|v| v / 1000.0),
        edge_c:      read_hwmon("amdgpu",  "temp1_input").map(|v| v / 1000.0),
        pkg_power_w: read_hwmon("amdgpu",  "power1_average").map(|v| v / 1_000_000.0),
        ec_temp_c: temp,
        ec_power_mode: mode,
        ec_fan1_rpm: rpm1,
        ec_fan2_rpm: rpm2,
        ec_fan3_rpm: rpm3,
    };
    Ok(serde_json::to_string(&s)?)
}

fn hostname() -> String {
    fs::read_to_string("/proc/sys/kernel/hostname")
        .map(|s| s.trim().to_string())
        .unwrap_or_else(|_| "unknown".into())
}

/// Walk /sys/class/hwmon/hwmon*/name and, if it matches `want`, read `file`
/// from that directory as a decimal integer.
fn read_hwmon(want: &str, file: &str) -> Option<f32> {
    let root = Path::new("/sys/class/hwmon");
    for entry in fs::read_dir(root).ok()?.flatten() {
        let name = fs::read_to_string(entry.path().join("name")).ok()?;
        if name.trim() == want {
            let raw = fs::read_to_string(entry.path().join(file)).ok()?;
            return raw.trim().parse::<f32>().ok();
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sample_is_valid_json() {
        // Missing hwmon files on a non-Strix box should still yield a
        // parseable line — everything just becomes null.
        let line = sample().unwrap();
        let v: serde_json::Value = serde_json::from_str(&line).unwrap();
        assert!(v.get("ts_unix").is_some());
        assert!(v.get("host").is_some());
    }
}

File: cpp/power/src/metrics.cpp
#include "onebit/power/metrics.hpp"

#include <chrono>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <system_error>

#include <nlohmann/json.hpp>

namespace onebit::power {

namespace {

[[nodiscard]] std::string trim_copy(std::string s)
{
    while (!s.empty() && (s.back() == '\n' || s.back() == '\r'
                       || s.back() == ' '  || s.back() == '\t')) s.pop_back();
    std::size_t i = 0;
    while (i < s.size() && (s[i] == ' ' || s[i] == '\t')) ++i;
    return s.substr(i);
}

[[nodiscard]] std::optional<std::string> read_file(const std::filesystem::path& p)
{
    std::ifstream f{p};
    if (!f) return std::nullopt;
    std::ostringstream buf; buf << f.rdbuf();
    return trim_copy(buf.str());
}

} // namespace

std::string read_hostname()
{
    auto v = read_file("/proc/sys/kernel/hostname");
    return v.value_or("unknown");
}

std::optional<float> read_hwmon(std::string_view want, std::string_view file)
{
    std::error_code ec;
    const std::filesystem::path root{"/sys/class/hwmon"};
    if (!std::filesystem::is_directory(root, ec)) return std::nullopt;

    for (const auto& entry : std::filesystem::directory_iterator(root, ec)) {
        if (ec) return std::nullopt;
        auto name = read_file(entry.path() / "name");
        if (!name || *name != want) continue;
        auto raw = read_file(entry.path() / file);
        if (!raw) return std::nullopt;
        try {
            return std::stof(*raw);
        } catch (...) {
            return std::nullopt;
        }
    }
    return std::nullopt;
}

Result<Sample> collect_sample(const EcBackend& ec)
{
    Sample s;
    s.ts_unix = static_cast<std::uint64_t>(
        std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch()).count());
    s.host = read_hostname();

    if (auto v = read_hwmon("k10temp", "temp1_input"))     s.tctl_c     = *v / 1000.0f;
    if (auto v = read_hwmon("amdgpu",  "temp1_input"))     s.edge_c     = *v / 1000.0f;
    if (auto v = read_hwmon("amdgpu",  "power1_average"))  s.pkg_power_w = *v / 1'000'000.0f;

    if (ec.available()) {
        if (auto r = ec.temp_c();     r) s.ec_temp_c     = r.value();
        if (auto r = ec.power_mode(); r) s.ec_power_mode = r.value();
        if (auto r = ec.fan(1);       r) s.ec_fan1_rpm   = r.value().rpm;
        if (auto r = ec.fan(2);       r) s.ec_fan2_rpm   = r.value().rpm;
        if (auto r = ec.fan(3);       r) s.ec_fan3_rpm   = r.value().rpm;
    }
    return s;
}

std::string sample_to_json(const Sample& s)
{
    // NB: must use parens, not braces — `nlohmann::json{x}` invokes the
    // initializer-list ctor and produces a 1-element ARRAY, not a scalar.
    auto opt_s = [](const std::optional<std::string>& v) -> nlohmann::json {
        return v ? nlohmann::json(*v) : nlohmann::json(nullptr);
    };
    auto opt_f = [](const std::optional<float>& v) -> nlohmann::json {
        return v ? nlohmann::json(*v) : nlohmann::json(nullptr);
    };
    auto opt_i = [](const std::optional<std::int32_t>& v) -> nlohmann::json {
        return v ? nlohmann::json(*v) : nlohmann::json(nullptr);
    };
    auto opt_u = [](const std::optional<std::uint32_t>& v) -> nlohmann::json {
        return v ? nlohmann::json(*v) : nlohmann::json(nullptr);
    };

    // Match Rust serde key order.
    nlohmann::ordered_json j;
    j["ts_unix"]       = s.ts_unix;
    j["host"]          = s.host;
    j["tctl_c"]        = opt_f(s.tctl_c);
    j["edge_c"]        = opt_f(s.edge_c);
    j["pkg_power_w"]   = opt_f(s.pkg_power_w);
    j["ec_temp_c"]     = opt_i(s.ec_temp_c);
    j["ec_power_mode"] = opt_s(s.ec_power_mode);
    j["ec_fan1_rpm"]   = opt_u(s.ec_fan1_rpm);
    j["ec_fan2_rpm"]   = opt_u(s.ec_fan2_rpm);
    j["ec_fan3_rpm"]   = opt_u(s.ec_fan3_rpm);
    return j.dump();
}

} // namespace onebit::power
